### ISSUE TITLE
samtools stats occasionally fails to read FASTA index file it generates

### ISSUE THREAD

**Issue by tneely** [samtools/samtools#1903]
#### Are you using the latest version of [samtools](https://github.com/samtools/samtools/releases/latest) and [HTSlib](https://github.com/samtools/htslib/releases/latest)? If not, please specify.

I am using samtools version `1.15.1`.

 #### Please describe your environment.
- OS `Linux 5.4.249-170.359.amzn2int.x86_64`
- machine architecture `x86_64`
- compiler `gcc (GCC) 11.3.1 20221121 (Red Hat 11.3.1-4)`

#### Please specify the steps taken to generate the issue, the command you are running and the relevant output.

I am running a canary that executes `samtools stats` over the same reference and BAM file every 10 minutes in a container on AWS ECS Fargate. I have seen over the days and weeks that this command occasionally fails with a `fai_read` error for some line of the reference index file. The exact line number it fails on is not consistent. Note that I am not providing the `.fai` file - samtools is automatically generating it as part of `stats` as part of each canary run. Because this runs in an ephemeral container, I do not have access to the generated FASTA index file, but I'll look into evacuating it on failure cases and update this issue when I hit another failure.

The failure rate is roughly 0.05-0.1%.

```
[E::fai_read] Could not understand FASTA index /reference.fasta.fai line 1051
[E::fai_load3_core] Failed to read FASTA index /reference.fasta.fai
Could not load faidx: /reference.fasta
```

Steps to reproduce:

Reference used : https://github.com/broadinstitute/gatk/blob/master/src/test/resources/large/Homo_sapiens_assembly38.fasta.gz
BAM file used: https://42basepairs.com/browse/s3/gatk-test-data/wgs_ubam/NA12878_20k#NA12878_A.bam

1. Download and unzip Homo_sapiens_assembly38.fasta.gz as reference.fasta
2. Download NA12878_A.bam
3. Run `samtools stats -@ 4 -r ./reference.fasta --reference ./reference.fasta ./NA12878_A.bam`

**Comment by daviesrob** [comment]
I'm not sure what would cause this, although I notice that your chosen BAM file is unaligned, so there's no reason to supply the reference at all.  There's also no reason to use the `--reference` option unless you're reading a CRAM file and there's no other way to get the references needed to uncompress it.

Leaving out the `--reference` option makes no difference to the `samtools stats` output on your file, apart from changing the reported command line.  Leaving out `-r` removes the `MPC` (mismatches per cycle) section, but for an unaligned file that's just full of zeros anyway.  You may want to consider finding an aligned bam file instead if you want to properly exercise the reference-handling code.

Meanwhile, if you do managed to catch a broken `.fai` file, it would be useful to compare it with a correctly-built one to find out what changed.

**Comment by jkbonfield** [comment]
> [E::fai_read] Could not understand FASTA index /reference.fasta.fai line 1051
> [E::fai_load3_core] Failed to read FASTA index /reference.fasta.fai
> Could not load faidx: /reference.fasta

This error comes from here: https://github.com/samtools/samtools/blob/develop/stats.c#L2453

```
            case 'r': info->fai = fai_load(optarg);
                      if (info->fai==NULL)
                          error("Could not load faidx: %s\n", optarg);
                      break;
```

So optarg here is "/references.fasta.fai".  Any clue why your fai file is supposedly in the root directory?  This is not the same directory nor filename as your quoted command line: `-r ./Homo_sapiens_assembly38.fasta`.

Short of argc/argv somehow becoming corrupted, which feels unlikely to do so in a manner to change filename like this, I suspect something else in your pipeline is running and mixing in errors from elsewhere as I simply cannot see this error message being produced from the command you list.

**Comment by tneely** [comment]
I was able to evacuate a couple of the generated index files associated with the `fai_read` error. They're both identical to each other and a separate FASTA index I had created locally:

```
$ diff evac1.fasta.fai Homo_sapiens_assembly38.fasta.fai
$ diff evac1.fasta.fai evac2.fasta.fai
$ diff Homo_sapiens_assembly38.fasta.fai evac2.fasta.fai
```

So the problem is in reading the file, not creating it.

**Comment by daviesrob** [comment]
If the problem is with reading the file, then it might be helpful to get `fai_read()` to print out what it read.  For example, you could add:
```
                fprintf(stderr, "%s\t%s\n", buf, p);
```
after [this hts_log_error() line](https://github.com/samtools/htslib/blob/6806b2d7f6bc892a66f26c2ace83ad3f7b1f50d9/faidx.c#L387).  Then let it run and see how the line as-read differs from the one in the known-good `.fai` file.

Apart from that, we'd need to know more about your workflow and how it's being run.  In particular, is there any chance that two processes could be trying to use the same reference at the same time?  That would offer a simple explanation for the issue you're seeing, if the second one tries to use a partially-written index made by the first.

**Comment by tneely** [comment]
>  In particular, is there any chance that two processes could be trying to use the same reference at the same time? That would offer a simple explanation for the issue you're seeing, if the second one tries to use a partially-written index made by the first.

Ah this is most likely the case. We're doing a few things in parallel with the BAM file, including `samtools stats` and `samtools view`, both of which are using the reference file. Thank you for the great insight here! We'll update the workflow on our end to create the faidx prior to these commands and see if that helps.

It's definitely sounding like the parallel command execution is the problem here, so I'll go ahead and close the issue. If this still isn't the case, I can reopen with further debugging logs from faidx.c.

**Comment by jmarshall** [comment]
There were previously cases in which downloading e.g. _.bai_ indexes in parallel would conflict with each other similarly. This was fixed in samtools/htslib#1112 by creating the index in a unique temporary filename and atomically renaming it to the final filename. At the time it didn't occur to me that *creating* an index as a byproduct of some command is a very similar case. It may be worth considering using `hts_open_tmpfile()` + `rename()` in this sort of case too.

**Comment by daviesrob** [comment]
@jmarshall That has crossed my mind, although in this case we'd need to ensure that the destination is a local filesystem so that `rename()` will work.
